http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2FrameExceptionHandler.java

```diff
@@ -0,0 +1,100 @@
+import io.netty.handler.timeout.ReadTimeoutException;
+import io.netty.handler.timeout.WriteTimeoutException;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.Logger;
+
+@ChannelHandler.Sharable
+@SdkInternalApi
+public final class Http2FrameExceptionHandler extends ChannelInboundHandlerAdapter {
+    private static final Http2FrameExceptionHandler INSTANCE = new Http2FrameExceptionHandler();
+    private static final Http2PingFrame DEFAULT_PING_FRAME = new DefaultHttp2PingFrame(0);
+    private static final Logger log = Logger.loggerFor(Http2FrameExceptionHandler.class);
+
+    private Http2FrameExceptionHandler() {
+    }
+
+    public static Http2FrameExceptionHandler create() {
+        return INSTANCE;
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        if ((cause instanceof ReadTimeoutException || cause instanceof WriteTimeoutException) && ctx.channel().parent() != null) {
+            healthCheckIfNotAlready(ctx);
+        }
+
+        ctx.fireExceptionCaught(cause);
+    }
+
+    private void healthCheckIfNotAlready(ChannelHandlerContext ctx) {
+        Channel parent = ctx.channel().parent();
+
+        // No need to do another health check if there is a PING in flight
+        if (PingStatus.IN_FLIGHT.equals(parent.attr(PING_STATUS).get())) {
+            return;
+        }
+
+        log.debug(() -> "Read/WriteTimeoutException occurred, sending PING frame to " + parent);
+        parent.attr(PING_STATUS).set(PingStatus.IN_FLIGHT);
+
+        parent.writeAndFlush(DEFAULT_PING_FRAME).addListener(res -> {
+            if (!res.isSuccess()) {
+                log.debug(() -> "Failed to write and flush PING frame to connection", res.cause());
+                closeH2Connection(parent);
+            }
+        });
+        parent.eventLoop().schedule(() -> timeoutHealthCheck(parent), HTTP2_CONNECTION_PING_TIMEOUT_SECONDS,
+                                TimeUnit.SECONDS);
+    }
+
+    private void timeoutHealthCheck(Channel parentChannel) {
+        if (!PingStatus.ACK_RECEIVED.equals(parentChannel.attr(PING_STATUS).get())) {
+            log.debug(() -> String.format("Has not received PING ACK within %s seconds, closing the connection %s",
+                                          HTTP2_CONNECTION_PING_TIMEOUT_SECONDS, parentChannel));
+            closeH2Connection(parentChannel);
+        }
+    }
+
+    private void closeH2Connection(Channel parent) {
+        MultiplexedChannelRecord channelRecord = parent.attr(CHANNEL_POOL_RECORD).get();
+        channelRecord.shutdownChildChannels(Http2FrameExceptionHandler.PingFailedException.INSTANCE);
+        parent.close();
+    }
+
+    static final class PingFailedException extends IOException {
+        static final PingFailedException INSTANCE = new PingFailedException("Failed to send PING to the service");
```

Comment:
Should our errors distinguish between failing to send, and failing to get a response?

Reply:
Sure